src/connector/src/state.rs

```diff
@@ -141,63 +92,29 @@ impl<S: StateStore> SourceStateHandler<S> {
     /// Initializes the state of the specified ``state_identifier`` and returns an empty vec
     /// if it does not exist (e.g., the first accessible source).
     ///
-    /// The function returns a collection of tuple (epoch, state).
-    pub async fn restore_states(&self, state_identifier: String) -> Result<Vec<(u64, Bytes)>> {
-        // TODO: do we need snapshot read here?
-        let epoch = u64::MAX;
-        let scan_rs = self.keyspace.scan(Option::None, epoch).await;
-        let mut restore_values = Vec::new();
-        match scan_rs {
-            Ok(scan_list) => {
-                for pair in scan_list {
-                    let stored_key = StateStoredKey::default().build_from_bytes(pair.0).unwrap();
-                    if stored_key
-                        .clone()
-                        .state_identifier
-                        .eq(&state_identifier.clone())
-                    {
-                        restore_values.push((stored_key.epoch, pair.1))
-                    }
-                }
-                Ok(restore_values)
-            }
-            Err(e) => Err(anyhow!(e)),
-        }
+    /// The function returns the serialized state.
+    pub async fn restore_states(
+        &self,
+        state_identifier: String,
+        epoch: u64,
+    ) -> Result<Option<Bytes>> {
+        self.keyspace.get(state_identifier, epoch).await.map_err(|e| anyhow!(e))
```

Comment:
according to #2546,  we should not write a state if there is no change so we may not find a state for some epochs. I suggest returning the state with the first epoch smaller or equal to the given epoch.

Reply:
The `get` semantics is exactly "return the state with the first epoch smaller or equal to the given epoch".